Tests/SentryProfilerTests/SentryProfilerSwiftTests.swift

```diff
@@ -84,31 +96,89 @@ class SentryProfilerSwiftTests: XCTestCase {
             }
 
     #if !os(macOS)
-            // gather mock GPU frame render timestamps
-            displayLinkWrapper.call() // call once directly to capture previous frame timestamp for comparison with later ones
-
-            var slowFrames = 0
-            var frozenFrames = 0
-            var frameRates = 0
-            for _ in 0..<max(mockSlowFramesPerBatch, mockFrozenFramesPerBatch, mockFrameRateChangesPerBatch.count) {
-                displayLinkWrapper.normalFrame()
-                if slowFrames < mockSlowFramesPerBatch {
-                    displayLinkWrapper.middlingSlowFrame()
-                    slowFrames += 1
+            var shouldRecordFrameRateExpectation = false
+            var _previousSystemTime = currentDateProvider.systemTime()
+
+            func changeFrameRate(_ new: FrameRate) {
+                displayLinkWrapper.changeFrameRate(new)
+                shouldRecordFrameRateExpectation = true
+            }
+
+            func renderGPUFrame(_ type: GPUFrame) {
+                if shouldRecordFrameRateExpectation {
+                    shouldRecordFrameRateExpectation = false
+                    print("will expect frame rate \(displayLinkWrapper.currentFrameRate.rawValue) at \(_previousSystemTime)")
+                    expectedFrameRateChanges.append([
+                        "elapsed_since_start_ns": String(_previousSystemTime),
+                        "value": NSNumber(value: displayLinkWrapper.currentFrameRate.rawValue)
+                    ])
                 }
-                displayLinkWrapper.normalFrame()
-                if frozenFrames < mockFrozenFramesPerBatch {
+
+                switch type {
+                case .normal:
+                    currentDateProvider.advanceBy(nanoseconds: 1)
```

Comment:
`l`: I think testing this with actual values that frames usually have would be better. 16.6 ms for a normal frame, slow frame 18 ms, and frozen frame 701 ms for 60 fps.

Reply:
Originally I did that, but it was tough to debug issues in the tests with lots of large, noninteger values.

Worth noting here that the clocks are still independent between `currentDateProvider` and `displayLinkWrapper`. I thought for a bit about unifying them for #2849 but since it technically works as-is, where `displayLinkWrapper` essentially performs the necessary modifications to cause a frame of the desired type to be recorded, I'm hesitant to modify it this go around. I agree it's confusing though, let me think about it a bit more today.

Also note that the profiling code that consumes these timestamps doesn't actually validate things like "this frame took X time to render, but its timestamp is only (X-y) time since the previous frame was rendered."

Reply:
Wound up doing this in #2856